cpmpy/expressions/globalfunctions.py

```diff
@@ -285,8 +285,14 @@ def decompose_comparison(self, cpm_op, cpm_rhs):
 
         """
         arr, idx = self.args
-        return [(idx == i).implies(eval_comparison(cpm_op, arr[i], cpm_rhs)) for i in range(len(arr))] + \
-               [idx >= 0, idx < len(arr)], []
+        # for every  `i` in the intersection of array indices and the bounds of `idx`..
+        lb, ub = get_bounds(idx)
+        new_lb, new_ub = max(lb, 0), min(ub, len(arr) - 1)
+        return [  # ..we post `(idx = i) -> idx=i -> arr[i] <CMP_OP> cpm_rhs`.
```

Comment:
I very very very much prefer the previous code with the explicit for-loop here... This is quite hard to parse and it is very unclear what is happening...

Reply:
I can change it back to that loop style (I'm a big fan of generators/maps/functional style, but it's a choice). Note though that this is how it was written before, just with a tighter loop, saver `implies`, and comments!

Reply:
Yes, I agree the generators are great, but not if they get so complicated we need to split them and add a comment to explain what they are doing ;)

Reply:
Maybe I could have structured it better (not concatenate them in place, the comments are a bit creatively placed). I will change it to a loop, I can definitely understand why this is harder to read. Although I like this because it just exactly specifies the decomp as you would mathematically: `idx=i -> arr[i] CMP rhs, for all i in range`.

Reply:
Ok, sounds good indeed!